aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -39,94 +50,418 @@ fn is_string_type(context: &SafeNativeContext, type_arg: &Type) -> SafeNativeRes
+/// fields: (`aggregator id`, `limit`).
+pub fn get_aggregator_fields_by_type(
+    ty_arg: &Type,
+    agg: &StructRef,
+) -> SafeNativeResult<(u128, u128)> {
+    match ty_arg {
+        Type::U128 => {
+            // Get aggregator information and a value to add.
+            let (id, limit) = get_aggregator_fields_u128(agg)?;
+            Ok((id, limit))
+        },
+        Type::U64 => {
+            // Get aggregator information and a value to add.
+            let (id, limit) = get_aggregator_fields_u64(agg)?;
+            Ok((id as u128, limit as u128))
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
+        }),
+    }
+}
+
+/// Given the list of native function arguments and a type, pop the next argument if it is of given type.
+pub fn pop_value_by_type(ty_arg: &Type, args: &mut VecDeque<Value>) -> SafeNativeResult<u128> {
+    match ty_arg {
+        Type::U128 => Ok(safely_pop_arg!(args, u128)),
+        Type::U64 => Ok(safely_pop_arg!(args, u64) as u128),
+        _ => Err(SafeNativeError::Abort {
+            abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
+        }),
+    }
+}
+
+pub fn create_value_by_type(ty_arg: &Type, value: u128) -> SafeNativeResult<Value> {
+    match ty_arg {
+        Type::U128 => Ok(Value::u128(value)),
+        Type::U64 => Ok(Value::u64(u128_to_u64(value)?)),
+        _ => Err(SafeNativeError::Abort {
+            abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
+        }),
+    }
+}
+
+// To avoid checking is_string_type multiple times, check type_arg only once, and convert into this enum
+enum SnapshotType {
+    U128,
+    U64,
+    String,
+}
+
+impl SnapshotType {
+    fn from_ty_arg(context: &SafeNativeContext, ty_arg: &Type) -> SafeNativeResult<Self> {
+        match ty_arg {
+            Type::U128 => Ok(Self::U128),
+            Type::U64 => Ok(Self::U64),
+            _ => {
+                // Check if the type is a string
+                if is_string_type(context, ty_arg)? {
+                    Ok(Self::String)
+                } else {
+                    // If not a string, return an error
+                    Err(SafeNativeError::Abort {
+                        abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+                    })
+                }
+            },
+        }
+    }
+
+    pub fn pop_snapshot_field_by_type(
+        &self,
+        args: &mut VecDeque<Value>,
+    ) -> SafeNativeResult<SnapshotValue> {
+        self.parse_snapshot_value_by_type(aggregator_snapshot_field_value(&safely_pop_arg!(
+            args, StructRef
+        ))?)
+    }
+
+    pub fn pop_snapshot_value_by_type(
+        &self,
+        args: &mut VecDeque<Value>,
+    ) -> SafeNativeResult<SnapshotValue> {
+        match self {
+            SnapshotType::U128 => Ok(SnapshotValue::Integer(safely_pop_arg!(args, u128))),
+            SnapshotType::U64 => Ok(SnapshotValue::Integer(safely_pop_arg!(args, u64) as u128)),
```

Comment:
Note for the future: SnapshotValue is replaced by DelayedFieldValue right?

Reply:
no - SnapshotValue encompasses types that AggregatorSnapshot can take


DelayedFieldValue is list of Delayed Fields (today aggregator, snapshot delta and snapshot derived. tomorrow maybe more).

this code is identical on aggregators_v2 branch - https://github.com/aptos-labs/aptos-core/blob/aggregators_v2/aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs#L143